File: backend/clarke_wright.py
"""
clarke_wright.py – Clarke-Wright Savings Algorithm for geo-aware vehicle assignment.

Replaces greedy bin-packing (FFD) with distance-optimised route clustering.

Key properties:
  - Minimises total driving distance by merging stops with highest "savings"
  - Uses fewest vehicles possible (consolidation-first)
  - Respects vehicle capacity strictly
  - Handles 'full' unit stops (Mega/Giga) — isolated, never merged
  - Graceful fallback when stops lack coordinates

Algorithm steps:
  1. Separate full-vehicle and no-coordinate stops (handled specially)
  2. Compute savings: saving(i,j) = d(depot→i) + d(depot→j) − d(i→j)
  3. Initialize each stop as its own route
  4. Greedily merge pairs with highest saving (respecting capacity)
  5. Assign merged routes to vehicles via best-fit decreasing
  6. Append no-coordinate and overflow stops to emptiest vehicles

Usage:
    from clarke_wright import clarke_wright_assign, has_coordinates

    assignments, unassigned = clarke_wright_assign(
        delivery_stops, vehicles, depot_coords=(55.516, 22.721)
    )
"""

from __future__ import annotations
import math

# Pagramantis, Tauragės r. – fixed depot coordinates
DEPOT_LAT = 55.516
DEPOT_LNG = 22.721

# Items whose name contains these keywords occupy the whole vehicle
FULL_VEHICLE_KEYWORDS = ("mega", "giga")

# Minimum fraction of stops that must have coordinates for CW to engage
MIN_COORD_FRACTION = 0.5


# ── Helpers ───────────────────────────────────────────────────────────────────

def _haversine(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    """Straight-line distance in km (Haversine formula)."""
    R = 6371.0
    phi1, phi2 = math.radians(lat1), math.radians(lat2)
    dphi = math.radians(lat2 - lat1)
    dlam = math.radians(lon2 - lon1)
    a = math.sin(dphi / 2) ** 2 + math.cos(phi1) * math.cos(phi2) * math.sin(dlam / 2) ** 2
    return R * 2 * math.asin(math.sqrt(a))


def _stop_dist(a: dict, b: dict) -> float:
    """Distance between two stops in km; inf if either lacks coordinates."""
    la, lo = a.get("lat"), a.get("lng")
    lb, lo2 = b.get("lat"), b.get("lng")
    if la is None or lo is None or lb is None or lo2 is None:
        return float("inf")
    return _haversine(float(la), float(lo), float(lb), float(lo2))


def _depot_dist(stop: dict, depot: tuple[float, float]) -> float:
    """Distance from depot to stop in km; 0 if coordinates missing."""
    la, lo = stop.get("lat"), stop.get("lng")
    if la is None or lo is None:
        return 0.0
    return _haversine(depot[0], depot[1], float(la), float(lo))


def _is_full(units) -> bool:
    return units == "full"


def _units_as_float(units, vehicle_capacity: float = 999.0) -> float:
    if units == "full":
        return float(vehicle_capacity)
    try:
        return float(units)
    except (TypeError, ValueError):
        return 1.0


def has_coordinates(stops: list[dict]) -> bool:
    """Return True when enough stops have lat/lng for Clarke-Wright to be useful."""
    if not stops:
        return False
    n_with = sum(1 for s in stops if s.get("lat") is not None and s.get("lng") is not None)
    return n_with >= max(2, int(len(stops) * MIN_COORD_FRACTION))


# ── Main algorithm ────────────────────────────────────────────────────────────

def clarke_wright_assign(
    stops: list[dict],
    vehicles: list[dict],
    depot_coords: tuple[float, float] = (DEPOT_LAT, DEPOT_LNG),
) -> tuple[dict[str, list[str]], list[str]]:
    """
    Assign delivery stops to vehicles using the Clarke-Wright savings algorithm.

    Args:
        stops:          List of stop dicts with keys: id, units (float|'full'),
                        lat (float|None), lng (float|None)
        vehicles:       List of vehicle dicts with keys: id, capacity (float)
        depot_coords:   (lat, lng) of the starting depot

    Returns:
        assignments     {vehicle_id: [stop_id, ...]}
        unassigned      [stop_id, ...]  — could not be placed (capacity exceeded)
    """
    if not vehicles:
        return {}, [s["id"] for s in stops]
    if not stops:
        return {v["id"]: [] for v in vehicles}, []

    # ── Partition stops ───────────────────────────────────────────────────────
    full_stops   = [s for s in stops if _is_full(s.get("units"))]
    reg_stops    = [s for s in stops if not _is_full(s.get("units"))]
    no_coord     = [s for s in reg_stops if s.get("lat") is None or s.get("lng") is None]
    coord_stops  = [s for s in reg_stops if s.get("lat") is not None and s.get("lng") is not None]

    n = len(coord_stops)

    # ── Sort vehicles: largest capacity first ─────────────────────────────────
    sorted_veh = sorted(vehicles, key=lambda v: -float(v.get("capacity", 4)))
    max_cap    = float(sorted_veh[0].get("capacity", 4)) if sorted_veh else 4.0

    # ── Compute savings matrix ────────────────────────────────────────────────
    # saving(i,j) = d(depot,i) + d(depot,j) - d(i,j)
    # High saving → stops are far from depot but close to each other → merge them
    savings: list[tuple[float, float, int, int]] = []
    for i in range(n):
        for j in range(i + 1, n):
            si, sj = coord_stops[i], coord_stops[j]
            d_di = _depot_dist(si, depot_coords)
            d_dj = _depot_dist(sj, depot_coords)
            d_ij = _stop_dist(si, sj)
            saving = d_di + d_dj - d_ij
            combined = (
                _units_as_float(si.get("units", 1))
                + _units_as_float(sj.get("units", 1))
            )
            # Tie-break: prefer pairs with more total units (consolidate heavy loads first)
            savings.append((saving, combined, i, j))

    savings.sort(key=lambda x: (-x[0], -x[1]))

    # ── Initialise: each stop is its own route ────────────────────────────────
    # route_id → ordered list of stop indices
    routes: dict[int, list[int]] = {i: [i] for i in range(n)}
    # stop index → current route id
    stop_route: dict[int, int] = {i: i for i in range(n)}
    # route_id → total units consumed
    route_units: dict[int, float] = {
        i: _units_as_float(coord_stops[i].get("units", 1)) for i in range(n)
    }
    next_rid = n  # monotonically increasing route id counter

    # ── Greedy merge loop ─────────────────────────────────────────────────────
    for _saving, _combined, i, j in savings:
        ri = stop_route.get(i)
        rj = stop_route.get(j)
        if ri is None or rj is None or ri == rj:
            continue

        route_i = routes[ri]
        route_j = routes[rj]

        # Both stops must be route endpoints (not interior nodes)
        if i not in (route_i[0], route_i[-1]):
            continue
        if j not in (route_j[0], route_j[-1]):
            continue

        # Merged capacity must not exceed the LARGEST available vehicle
        # (we'll do best-fit assignment to actual vehicles afterwards)
        merged_u = route_units[ri] + route_units[rj]
        if merged_u > max_cap + 0.01:
            continue

        # Orient: i at tail of route_i, j at head of route_j, then concatenate
        if route_i[0] == i:
            route_i = list(reversed(route_i))
        if route_j[-1] == j:
            route_j = list(reversed(route_j))

        merged = route_i + route_j

        # Create new merged route
        routes[next_rid]     = merged
        route_units[next_rid] = merged_u
        for idx in merged:
            stop_route[idx] = next_rid
        del routes[ri], routes[rj]
        next_rid += 1

    # ── Collect final routes (deduplicated) ───────────────────────────────────
    seen: set[int] = set()
    final_routes: list[tuple[float, list[str]]] = []   # (total_units, [stop_id, ...])
    for idx in range(n):
        rid = stop_route[idx]
        if rid in seen:
            continue
        seen.add(rid)
        stop_ids = [coord_stops[k]["id"] for k in routes[rid]]
        final_routes.append((route_units[rid], stop_ids))

    # Largest route first → gets assigned to largest vehicle
    final_routes.sort(key=lambda x: -x[0])

    # ── Assign to vehicles ────────────────────────────────────────────────────
    assignments: dict[str, list[str]]  = {v["id"]: [] for v in vehicles}
    remaining:   dict[str, float]      = {v["id"]: float(v.get("capacity", 4)) for v in vehicles}
    unassigned:  list[str]             = []

    # 1. Full-vehicle stops first — each needs an empty vehicle
    for fs in full_stops:
        placed = False
        for v in sorted_veh:
            vid = v["id"]
            cap = float(v.get("capacity", 4))
            if abs(remaining[vid] - cap) < 0.01:   # still empty
                assignments[vid].append(fs["id"])
                remaining[vid] = 0.0
                placed = True
                break
        if not placed:
            unassigned.append(fs["id"])

    # 2. Clarke-Wright routes — best-fit decreasing into remaining vehicles
    for r_units, stop_ids in final_routes:
        # Best fit: vehicle where (remaining − route_units) is smallest non-negative
        best_vid   = None
        best_slack = float("inf")
        for v in sorted_veh:
            vid = v["id"]
            slack = remaining[vid] - r_units
            if slack >= -0.01 and slack < best_slack:
                best_slack = slack
                best_vid   = vid

        if best_vid is not None:
            assignments[best_vid].extend(stop_ids)
            remaining[best_vid] = round(remaining[best_vid] - r_units, 4)
        else:
            # Route is too large for any vehicle — split and assign individually
            for sid in stop_ids:
                stop_obj  = next((s for s in coord_stops if s["id"] == sid), None)
                su        = _units_as_float(stop_obj.get("units", 1) if stop_obj else 1)
                placed    = False
                for v in sorted_veh:
                    vid = v["id"]
                    if remaining[vid] >= su - 0.01:
                        assignments[vid].append(sid)
                        remaining[vid] = round(remaining[vid] - su, 4)
                        placed = True
                        break
                if not placed:
                    unassigned.append(sid)

    # 3. No-coordinate stops — assign to vehicle with most remaining capacity
    for ns in no_coord:
        if not sorted_veh:
            unassigned.append(ns["id"])
            continue
        best_vid = max(remaining, key=remaining.get)
        best_v   = next(v for v in vehicles if v["id"] == best_vid)
        su       = _units_as_float(ns.get("units", 1), float(best_v.get("capacity", 4)))
        if remaining[best_vid] >= su - 0.01:
            assignments[best_vid].append(ns["id"])
            remaining[best_vid] = round(remaining[best_vid] - su, 4)
        else:
            # Try remaining vehicles
            placed = False
            for v in sorted_veh:
                vid = v["id"]
                su2 = _units_as_float(ns.get("units", 1), float(v.get("capacity", 4)))
                if remaining[vid] >= su2 - 0.01:
                    assignments[vid].append(ns["id"])
                    remaining[vid] = round(remaining[vid] - su2, 4)
                    placed = True
                    break
            if not placed:
                unassigned.append(ns["id"])

    return assignments, unassigned

File: frontend/src/utils/clarkeWright.js
/**
 * clarkeWright.js – Clarke-Wright Savings Algorithm (JS port of backend/clarke_wright.py)
 *
 * Running vehicle assignment in the browser means backend code never needs to
 * change when the algorithm is updated — push to git = deployed via GitHub Pages.
 *
 * Also includes binPackAssign() as a fallback when stops lack coordinates,
 * and assignPickups() to mirror the pickup-matching logic from server.py.
 *
 * Exports:
 *   hasCoordinates(stops)                           → boolean
 *   clarkeWrightAssign(stops, vehicles)             → { assignments, unassigned }
 *   binPackAssign(stops, vehicles)                  → { assignments, unassigned }
 *   assignPickups(pickupStops, assignments,
 *                 deliveryStops, vehicles)           → { assignments, unassigned }
 */

// Pagramantis, Tauragės r. – fixed depot coordinates
const DEPOT_LAT = 55.516;
const DEPOT_LNG = 22.721;

// Minimum fraction of stops that must have coordinates for CW to engage
const MIN_COORD_FRACTION = 0.5;


// ── Helpers ───────────────────────────────────────────────────────────────────

function haversine(lat1, lon1, lat2, lon2) {
  const R = 6371.0;
  const toRad = deg => (deg * Math.PI) / 180;
  const phi1 = toRad(lat1);
  const phi2 = toRad(lat2);
  const dPhi = toRad(lat2 - lat1);
  const dLam = toRad(lon2 - lon1);
  const a =
    Math.sin(dPhi / 2) ** 2 +
    Math.cos(phi1) * Math.cos(phi2) * Math.sin(dLam / 2) ** 2;
  return R * 2 * Math.asin(Math.sqrt(a));
}

function stopDist(a, b) {
  const la = a.lat, lo = a.lng;
  const lb = b.lat, lo2 = b.lng;
  if (la == null || lo == null || lb == null || lo2 == null) return Infinity;
  return haversine(Number(la), Number(lo), Number(lb), Number(lo2));
}

function depotDist(stop, depot) {
  const la = stop.lat, lo = stop.lng;
  if (la == null || lo == null) return 0;
  return haversine(depot[0], depot[1], Number(la), Number(lo));
}

function isFullUnit(units) {
  return units === 'full';
}

function unitsAsFloat(units, vehicleCapacity = 999) {
  if (units === 'full') return Number(vehicleCapacity);
  const n = Number(units);
  return isNaN(n) ? 1.0 : n;
}


// ── hasCoordinates ─────────────────────────────────────────────────────────────

/**
 * Return true when enough stops have lat/lng for Clarke-Wright to be useful.
 * Mirrors: has_coordinates() in clarke_wright.py
 */
export function hasCoordinates(stops) {
  if (!stops || stops.length === 0) return false;
  const nWith = stops.filter(s => s.lat != null && s.lng != null).length;
  return nWith >= Math.max(2, Math.floor(stops.length * MIN_COORD_FRACTION));
}


// ── clarkeWrightAssign ────────────────────────────────────────────────────────

/**
 * Assign delivery stops to vehicles using Clarke-Wright savings algorithm.
 * Mirrors: clarke_wright_assign() in clarke_wright.py
 *
 * @param {Array}  stops    - [{id, units, lat, lng, ...}]
 * @param {Array}  vehicles - [{id, capacity, ...}]
 * @returns {{ assignments: Object, unassigned: Array }}
 *   assignments = { vehicleId: [stopId, ...] }
 *   unassigned  = [stopId, ...]
 */
export function clarkeWrightAssign(
  stops,
  vehicles,
  depotCoords = [DEPOT_LAT, DEPOT_LNG],
) {
  if (!vehicles || vehicles.length === 0) {
    return { assignments: {}, unassigned: stops.map(s => s.id) };
  }
  if (!stops || stops.length === 0) {
    const assignments = {};
    vehicles.forEach(v => { assignments[v.id] = []; });
    return { assignments, unassigned: [] };
  }

  // ── Partition stops ─────────────────────────────────────────────────────────
  const fullStops  = stops.filter(s => isFullUnit(s.units));
  const regStops   = stops.filter(s => !isFullUnit(s.units));
  const noCoord    = regStops.filter(s => s.lat == null || s.lng == null);
  const coordStops = regStops.filter(s => s.lat != null && s.lng != null);
  const n          = coordStops.length;

  // ── Sort vehicles: largest capacity first ───────────────────────────────────
  const sortedVeh = [...vehicles].sort(
    (a, b) => Number(b.capacity || 4) - Number(a.capacity || 4),
  );
  const maxCap = sortedVeh.length > 0 ? Number(sortedVeh[0].capacity || 4) : 4;

  // ── Compute savings matrix ──────────────────────────────────────────────────
  // saving(i,j) = d(depot,i) + d(depot,j) - d(i,j)
  const savings = [];
  for (let i = 0; i < n; i++) {
    for (let j = i + 1; j < n; j++) {
      const si = coordStops[i];
      const sj = coordStops[j];
      const dDi = depotDist(si, depotCoords);
      const dDj = depotDist(sj, depotCoords);
      const dIj = stopDist(si, sj);
      const saving   = dDi + dDj - dIj;
      const combined = unitsAsFloat(si.units ?? 1) + unitsAsFloat(sj.units ?? 1);
      savings.push([saving, combined, i, j]);
    }
  }
  // Sort descending by saving, tie-break by combined units
  savings.sort((a, b) => b[0] - a[0] || b[1] - a[1]);

  // ── Initialise: each stop is its own route ──────────────────────────────────
  // routes: routeId → [stopIndex, ...]
  const routes     = {};
  const stopRoute  = {};    // stopIndex → routeId
  const routeUnits = {};    // routeId   → total units
  for (let i = 0; i < n; i++) {
    routes[i]     = [i];
    stopRoute[i]  = i;
    routeUnits[i] = unitsAsFloat(coordStops[i].units ?? 1);
  }
  let nextRid = n;

  // ── Greedy merge loop ───────────────────────────────────────────────────────
  for (const [, , i, j] of savings) {
    const ri = stopRoute[i];
    const rj = stopRoute[j];
    if (ri == null || rj == null || ri === rj) continue;

    let routeI = routes[ri];
    let routeJ = routes[rj];

    // Both stops must be route endpoints
    if (routeI[0] !== i && routeI[routeI.length - 1] !== i) continue;
    if (routeJ[0] !== j && routeJ[routeJ.length - 1] !== j) continue;

    // Capacity check
    const mergedU = routeUnits[ri] + routeUnits[rj];
    if (mergedU > maxCap + 0.01) continue;

    // Orient: i at tail of routeI, j at head of routeJ
    if (routeI[0] === i) routeI = [...routeI].reverse();
    if (routeJ[routeJ.length - 1] === j) routeJ = [...routeJ].reverse();

    const merged = [...routeI, ...routeJ];

    routes[nextRid]     = merged;
    routeUnits[nextRid] = mergedU;
    merged.forEach(idx => { stopRoute[idx] = nextRid; });
    delete routes[ri];
    delete routes[rj];
    nextRid++;
  }

  // ── Collect final routes ────────────────────────────────────────────────────
  const seen         = new Set();
  const finalRoutes  = [];  // [{ units, stopIds }]
  for (let idx = 0; idx < n; idx++) {
    const rid = stopRoute[idx];
    if (seen.has(rid)) continue;
    seen.add(rid);
    finalRoutes.push({
      units:   routeUnits[rid],
      stopIds: routes[rid].map(k => coordStops[k].id),
    });
  }
  finalRoutes.sort((a, b) => b.units - a.units);  // largest first

  // ── Assign to vehicles ──────────────────────────────────────────────────────
  const assignments = {};
  const remaining   = {};
  vehicles.forEach(v => {
    assignments[v.id] = [];
    remaining[v.id]   = Number(v.capacity || 4);
  });
  const unassigned = [];

  // 1. Full-vehicle stops → each needs an empty vehicle
  for (const fs of fullStops) {
    let placed = false;
    for (const v of sortedVeh) {
      const cap = Number(v.capacity || 4);
      if (Math.abs(remaining[v.id] - cap) < 0.01) {  // still empty
        assignments[v.id].push(fs.id);
        remaining[v.id] = 0;
        placed = true;
        break;
      }
    }
    if (!placed) unassigned.push(fs.id);
  }

  // 2. CW routes → first-fit decreasing: assign to the first (largest) vehicle
  // that can hold this route. sortedVeh is already sorted largest-capacity-first,
  // so this guarantees big vehicles fill up before small ones.
  for (const { units: rUnits, stopIds } of finalRoutes) {
    let bestVid = null;
    for (const v of sortedVeh) {
      if (remaining[v.id] >= rUnits - 0.01) {
        bestVid = v.id;
        break;  // first fit: stop at the first (largest) vehicle that fits
      }
    }

    if (bestVid !== null) {
      assignments[bestVid].push(...stopIds);
      remaining[bestVid] = Math.round((remaining[bestVid] - rUnits) * 10000) / 10000;
    } else {
      // Route too large — assign individually
      for (const sid of stopIds) {
        const stopObj = coordStops.find(s => s.id === sid);
        const su      = unitsAsFloat(stopObj?.units ?? 1);
        let placed    = false;
        for (const v of sortedVeh) {
          if (remaining[v.id] >= su - 0.01) {
            assignments[v.id].push(sid);
            remaining[v.id] = Math.round((remaining[v.id] - su) * 10000) / 10000;
            placed = true;
            break;
          }
        }
        if (!placed) unassigned.push(sid);
      }
    }
  }

  // 3. No-coordinate stops → first-fit (largest vehicle first).
  // sortedVeh is already sorted largest-capacity-first.
  for (const ns of noCoord) {
    if (sortedVeh.length === 0) { unassigned.push(ns.id); continue; }

    let placed = false;
    for (const v of sortedVeh) {
      const su = unitsAsFloat(ns.units ?? 1, Number(v.capacity || 4));
      if (remaining[v.id] >= su - 0.01) {
        assignments[v.id].push(ns.id);
        remaining[v.id] = Math.round((remaining[v.id] - su) * 10000) / 10000;
        placed = true;
        break;
      }
    }
    if (!placed) unassigned.push(ns.id);
  }

  return { assignments, unassigned };
}


// ── binPackAssign ─────────────────────────────────────────────────────────────

/**
 * Greedy bin-packing fallback (used when stops lack coordinates).
 * Mirrors: bin_pack() in bin_pack.py
 *
 * @param {Array}  stops    - [{id, units, ...}]
 * @param {Array}  vehicles - [{id, capacity, ...}]
 * @returns {{ assignments: Object, unassigned: Array }}
 */
export function binPackAssign(stops, vehicles) {
  if (!vehicles || vehicles.length === 0) {
    return { assignments: {}, unassigned: stops.map(s => s.id) };
  }

  // Sort: 'full' first, then by units descending
  const sorted = [...stops].sort((a, b) => {
    const au = a.units, bu = b.units;
    if (au === 'full' && bu !== 'full') return -1;
    if (bu === 'full' && au !== 'full') return  1;
    return unitsAsFloat(bu ?? 1) - unitsAsFloat(au ?? 1);
  });

  // Sort vehicles largest-first so first-fit-decreasing fills large vehicles first
  const sortedVehicles = [...vehicles].sort(
    (a, b) => Number(b.capacity || 4) - Number(a.capacity || 4),
  );

  const remaining   = {};
  const assignments = {};
  sortedVehicles.forEach(v => {
    remaining[v.id]   = Number(v.capacity || 4);
    assignments[v.id] = [];
  });
  // Ensure all original vehicles have an entry even if capacity is 0
  vehicles.forEach(v => {
    if (!(v.id in assignments)) assignments[v.id] = [];
  });
  const unassigned = [];

  for (const stop of sorted) {
    const rawUnits = stop.units;
    let placed = false;

    for (const v of sortedVehicles) {
      const cap = Number(v.capacity || 4);
      const rem = remaining[v.id];
      const eff = unitsAsFloat(rawUnits, cap);

      if (rawUnits === 'full') {
        if (Math.abs(rem - cap) < 0.01) {  // vehicle is empty
          assignments[v.id].push(stop.id);
          remaining[v.id] = 0;
          placed = true;
          break;
        }
      } else {
        if (rem >= eff - 0.01) {
          assignments[v.id].push(stop.id);
          remaining[v.id] = Math.round((rem - eff) * 10000) / 10000;
          placed = true;
          break;
        }
      }
    }

    if (!placed) unassigned.push(stop.id);
  }

  return { assignments, unassigned };
}


// ── assignPickups ─────────────────────────────────────────────────────────────

/**
 * Assign pickup stops to the same vehicle as their delivery counterpart.
 * Matches by formattedAddress (same address = same event = same vehicle).
 * Mirrors the pickup-matching logic in optimize_route_multi (server.py).
 *
 * @param {Array}  pickupStops  - stops with type !== 'delivery'
 * @param {Object} assignments  - current assignments from CW/bin-pack (delivery only)
 * @param {Array}  deliveryStops - all delivery stops (for address lookup)
 * @param {Array}  vehicles      - [{id, ...}]
 * @returns {{ assignments: Object, unassigned: Array }}
 *   Returns a NEW assignments object (does not mutate input).
 */
export function assignPickups(pickupStops, assignments, deliveryStops, vehicles) {
  // Deep-copy assignments so we don't mutate the input
  const result = {};
  Object.entries(assignments).forEach(([vid, ids]) => {
    result[vid] = [...ids];
  });

  // Build address → vehicleId map from delivery assignments
  const addrToVehicle = {};
  Object.entries(result).forEach(([vid, sids]) => {
    sids.forEach(sid => {
      const s = deliveryStops.find(x => x.id === sid);
      if (s) {
        const addr = (s.formattedAddress || s.address || '').trim().toLowerCase();
        if (addr) addrToVehicle[addr] = vid;
      }
    });
  });

  const unassigned = [];
  for (const ps of pickupStops) {
    const addr      = (ps.formattedAddress || ps.address || '').trim().toLowerCase();
    const targetVid = addrToVehicle[addr];

    if (targetVid && result[targetVid] !== undefined) {
      result[targetVid].push(ps.id);
    } else if (vehicles && vehicles.length > 0) {
      // Fallback: assign to the largest vehicle (big cars get priority)
      const largestVid = [...vehicles].sort(
        (a, b) => Number(b.capacity || 4) - Number(a.capacity || 4),
      )[0].id;
      result[largestVid].push(ps.id);
    } else {
      unassigned.push(ps.id);
    }
  }

  return { assignments: result, unassigned };
}
